quinn-udp/src/cmsg/mod.rs

```diff
@@ -0,0 +1,138 @@
+use std::{
+    ffi::{c_int, c_uchar},
+    mem, ptr,
+};
+
+#[cfg(unix)]
+#[path = "unix.rs"]
+mod imp;
+
+#[cfg(windows)]
+#[path = "windows.rs"]
+mod imp;
+
+pub(crate) use imp::Aligned;
+
+// Helper traits for native types for control messages
+pub(crate) trait MsgHdrHelper {
+    type Output;
```

Comment:
Maybe name this `ControlHeader` for clarity? Also, could we add a `: CMsgHdrHelper` constraint to it? That should allow the rest of the code in this module to have only a single type variable rather than two.

Reply:
An associated type is always uniquely identified by the type it's associated with:
```diff
diff --git a/quinn-udp/src/cmsg/mod.rs b/quinn-udp/src/cmsg/mod.rs
index 78d449f4..cc5ecdc7 100644
--- a/quinn-udp/src/cmsg/mod.rs
+++ b/quinn-udp/src/cmsg/mod.rs
@@ -15,7 +15,7 @@ pub(crate) use imp::Aligned;
 
 // Helper traits for native types for control messages
 pub(crate) trait MsgHdr {
-    type ControlMessage;
+    type ControlMessage: CMsgHdr;
 
     fn control_len(&self) -> usize;
 
@@ -43,13 +43,13 @@ pub(crate) trait CMsgHdr {
 ///
 /// The operation must be "finished" for the native msghdr to be usable, either by calling `finish`
 /// explicitly or by dropping the `Encoder`.
-pub(crate) struct Encoder<'a, M: MsgHdr, C: CMsgHdr> {
+pub(crate) struct Encoder<'a, M: MsgHdr> {
     hdr: &'a mut M,
-    cmsg: Option<&'a mut C>,
+    cmsg: Option<&'a mut M::ControlMessage>,
     len: usize,
 }
 
-impl<'a, M: MsgHdr<ControlMessage = C>, C: CMsgHdr> Encoder<'a, M, C> {
+impl<'a, M: MsgHdr> Encoder<'a, M> {
     /// # Safety
     /// - `hdr` must contains a suitably aligned pointer to a buffer big enought to hold the control messages bytes
     //    that can be safely written
@@ -68,8 +68,8 @@ impl<'a, M: MsgHdr<ControlMessage = C>, C: CMsgHdr> Encoder<'a, M, C> {
     /// - If insufficient buffer space remains.
     /// - If `T` has stricter alignment requirements than native `cmsghdr`
     pub(crate) fn push<T: Copy + ?Sized>(&mut self, level: c_int, ty: c_int, value: T) {
-        assert!(mem::align_of::<T>() <= mem::align_of::<C>());
-        let space = C::cmsg_space(mem::size_of_val(&value));
+        assert!(mem::align_of::<T>() <= mem::align_of::<M::ControlMessage>());
+        let space = M::ControlMessage::cmsg_space(mem::size_of_val(&value));
         assert!(
             self.hdr.control_len() >= self.len + space,
             "control message buffer too small. Required: {}, Available: {}",
@@ -77,7 +77,11 @@ impl<'a, M: MsgHdr<ControlMessage = C>, C: CMsgHdr> Encoder<'a, M, C> {
             self.hdr.control_len()
         );
         let cmsg = self.cmsg.take().expect("no control buffer space remaining");
-        cmsg.set(level, ty, C::cmsg_len(mem::size_of_val(&value)));
+        cmsg.set(
+            level,
+            ty,
+            M::ControlMessage::cmsg_len(mem::size_of_val(&value)),
+        );
         unsafe {
             ptr::write(cmsg.cmsg_data() as *const T as *mut T, value);
         }
@@ -93,7 +97,7 @@ impl<'a, M: MsgHdr<ControlMessage = C>, C: CMsgHdr> Encoder<'a, M, C> {
 
 // Statically guarantees that the encoding operation is "finished" before the control buffer is read
 // by `sendmsg` like API.
-impl<'a, M: MsgHdr, C: CMsgHdr> Drop for Encoder<'a, M, C> {
+impl<'a, M: MsgHdr> Drop for Encoder<'a, M> {
     fn drop(&mut self) {
         self.hdr.set_control_len(self.len as _);
     }
@@ -108,12 +112,12 @@ pub(crate) unsafe fn decode<T: Copy, C: CMsgHdr>(cmsg: &C) -> T {
     ptr::read(cmsg.cmsg_data() as *const T)
 }
 
-pub(crate) struct Iter<'a, M: MsgHdr, C: CMsgHdr> {
+pub(crate) struct Iter<'a, M: MsgHdr> {
     hdr: &'a M,
-    cmsg: Option<&'a C>,
+    cmsg: Option<&'a M::ControlMessage>,
 }
 
-impl<'a, M: MsgHdr<ControlMessage = C>, C: CMsgHdr> Iter<'a, M, C> {
+impl<'a, M: MsgHdr> Iter<'a, M> {
     /// # Safety
     ///
     /// `hdr` must hold a pointer to memory outliving `'a` which can be soundly read for the
@@ -127,8 +131,8 @@ impl<'a, M: MsgHdr<ControlMessage = C>, C: CMsgHdr> Iter<'a, M, C> {
     }
 }
 
-impl<'a, M: MsgHdr<ControlMessage = C>, C: CMsgHdr> Iterator for Iter<'a, M, C> {
-    type Item = &'a C;
+impl<'a, M: MsgHdr> Iterator for Iter<'a, M> {
+    type Item = &'a M::ControlMessage;
 
     fn next(&mut self) -> Option<Self::Item> {
         let current = self.cmsg.take()?;
diff --git a/quinn-udp/src/unix.rs b/quinn-udp/src/unix.rs
index 787679bb..d7d47814 100644
--- a/quinn-udp/src/unix.rs
+++ b/quinn-udp/src/unix.rs
@@ -770,10 +770,7 @@ mod gso {
         }
     }
 
-    pub(crate) fn set_segment_size(
-        encoder: &mut cmsg::Encoder<libc::msghdr, libc::cmsghdr>,
-        segment_size: u16,
-    ) {
+    pub(crate) fn set_segment_size(encoder: &mut cmsg::Encoder<libc::msghdr>, segment_size: u16) {
         encoder.push(libc::SOL_UDP, libc::UDP_SEGMENT, segment_size);
     }
 }
```